docs/en/writing-running-appium/android_coverage.md

```diff
@@ -0,0 +1,292 @@
+        activity.registerReceiver(broadcast, new IntentFilter("com.example.pkg.END_EMMA"));
+    }
+    private String getCoverageFilePath() {
+        if (mCoverageFilePath == null) {
+            return DEFAULT_COVERAGE_FILE_PATH;
+        } else {
+            return mCoverageFilePath;
+        }
+    }
+    private void generateCoverageReport() {
+        Log.d(TAG, "generateCoverageReport():" + getCoverageFilePath());
+        OutputStream out = null;
+        try {
+            out = new FileOutputStream(getCoverageFilePath(), false);
+            Object agent = Class.forName("org.jacoco.agent.rt.RT")
+                    .getMethod("getAgent")
+                    .invoke(null);
+            out.write((byte[]) agent.getClass().getMethod("getExecutionData", boolean.class)
+                    .invoke(agent, false));
+        } catch (Exception e) {
+            Log.d(TAG, e.toString(), e);
+        } finally {
+            if (out != null) {
+                try {
+                    out.close();
+                } catch (IOException e) {
+                    e.printStackTrace();
+                }
+            }
+        }
+    }
+
+    @Override
+    public void onActivityEnd() {
+        if (LOGD)      Log.d(TAG, "onActivityFinished()");
+        if (mCoverage) {
+            generateCoverageReport();
+        }
+        finish(Activity.RESULT_OK, mResults);
+    }
+}
+
+```
+
+Then, `InstrumentActivityListener.java`
+```java
+package com.example.pkg;
+
+public interface InstrumentActivityListener {
+    void onActivityEnd();
+}
+```
+
+`InstrumentActivity.java` (This is optional, you can use MainActivity)
+```java
+package com.example.pkg;
+import android.app.Instrumentation;
+import android.os.Bundle;
+import android.util.Log;
+
+
+public class InstrumentActivity extends MainActivity {
+    public static String TAG = "IntrumentedActivity";
+    private InstrumentActivityListener listener;
+
+    public void setInstrumentActivityListener(InstrumentActivityListener listener) {
+        this.listener = listener;
+    }
+
+    // Generate output report when the activity is destroyed
+    @Override
+    public void onDestroy() {
+        super.onDestroy();
+        Log.d(TAG, "onDestroy()");
+        super.finish();
+        if (listener != null) {
+            listener.onActivityEnd();
+        }
+    }
+}
+```
+
+Finally, the most important part is `gradle`:
+```groovy
+....
+
+
+apply plugin: 'jacoco' // add plugin for jacoco
+
+...
+
+android {
+    ...
+    defaultConfig {
+        ...
+        testInstrumentationRunner "android.support.test.runner.AndroidJUnitRunner"
+    }
+
+    buildTypes {
+        release {
+            ...
+        }
+        debug{
+            testCoverageEnabled = true
+        }
+    }
+}
+
+dependencies {
+    ...
+    //uiautomator
+    androidTestCompile 'com.android.support.test.uiautomator:uiautomator-v18:2.1.0'
+}
+
+```
+
+
+### [3] Now, build apk and run Appium tests!!
```

Comment:
`Now, build the apk and run Appium tests!` (one exclamation point should be enough, I think! even though this is indeed a very exciting situation)